lazy.go

```diff
@@ -51,7 +51,11 @@ type lazyConn struct {
 }
 
 func (l *lazyConn) Read(b []byte) (int, error) {
-	if !l.rhandshake {
+	l.rhlock.RLock()
+	rhandshake := l.rhandshake
+	l.rhlock.RUnlock()
```

Comment:
This check was intentionally racy to avoid atomic operations on the fast path. Go routines should erroneously conclude that the handshake hasn't happened at most once (and the synchronization inside the handshake functions should prevent double handshakes).

Reply:
We could have two Once's, one for the read handshake, one for the write handshake (where each handshake side would asynchronously fire off the other side of the handshake (in a once)).

Reply:
Also, sorry for the massive delay...